File: set2/6/Main.java
import java.io.*;
import java.util.*;
import java.lang.*;

public class Main {
    final static int days[] = {0,31,28,31,30,31,30,31,31,30,31,30,31};
    static public void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        String []line = br.readLine().split(" ");

        // Input
        int N = Integer.parseInt(line[0]);
        String []tmp = line[1].split("/");
        int day    = Integer.parseInt(tmp[0]);
        int hour   = Integer.parseInt(tmp[1].split(":")[0]);
        int minute = Integer.parseInt(tmp[1].split(":")[1]);
        long limit = (hour + day * 24) * 60 + minute;

        int money  = Integer.parseInt(line[2]);

        HashMap<String, HashMap<String, Long>> mp = new HashMap<>();
        HashMap<String, Long> fine = new HashMap<>();
        for(int i = 0; i < N; ++i) {
            line = br.readLine().split(" ");
            long curTime = getTime(line[0], line[1]);
            String nickname = line[3];
            String partname = line[2];
            
            if(mp.containsKey(nickname)) {
                if(mp.get(nickname).containsKey(partname)) {
                    long prev = mp.get(nickname).get(partname);
                    long dist = Math.max(0L, curTime - prev - limit);
                    
                    if(fine.containsKey(nickname)) {
                        long update = fine.get(nickname) + dist * money;
                        fine.put(nickname, update);
                    }
                    else {
                        fine.put(nickname, dist * money);
                    }

                    mp.get(nickname).remove(partname);
                }
                else {
                    mp.get(nickname).put(partname, curTime);
                }
            }
            else {
                mp.put(nickname, new HashMap<String, Long>());
                mp.get(nickname).put(partname, curTime);
            }
        }

        List<String> keyList = new ArrayList<>(fine.keySet());
        keyList.sort((s1, s2) -> s1.compareTo(s2));

        Boolean ans = false;
        StringBuilder out = new StringBuilder();
        for(String key: keyList) {
            if(fine.get(key) == 0)continue;
            out.append(key + " " + fine.get(key) + '\n');
            ans = true;
        }
        if(ans == false) {
            out.append("-1");
        }
        System.out.print(out);
    }

    static long getDate(String date) {
        String []line = date.split("-");
        int month = Integer.parseInt(line[1]);
        int day   = Integer.parseInt(line[2]);

        long ret = day - 1;
        for(int i = 0; i < month; ++i) ret += days[i];
        return ret;
    }

    static long getTime(String time) {
        String []line = time.split(":");
        int hh = Integer.parseInt(line[0]);
        int mm = Integer.parseInt(line[1]);
        return hh * 60 + mm;
    }

    static long getTime(String date, String time) {
        String []line = time.split(":");
        int hh = Integer.parseInt(line[0]);
        int mm = Integer.parseInt(line[1]);
        return (hh + getDate(date) * 24) * 60 + mm;
    }
}

File: set2/6/main.py
import sys

def input():
    return sys.stdin.readline().strip()

N, limit, money = input().split(" ")
N, money = int(N), int(money)
days = [0,31,28,31,30,31,30,31,31,30,31,30,31]

def getTime(time):
    hh, mm = map(int, time.split(':'))
    return hh * 60 + mm

def getDate(date):
    year, month, day = map(int, date.split('-'))
    ret = day - 1
    for i in range(month):
        ret += days[i]
    return ret

def getTime2(date, time):
    hh, mm = map(int, time.split(':'))
    return (hh + getDate(date) * 24) * 60 + mm

def getLine():
    date, time, partname, nickname = input().split()
    return getTime2(date, time), partname, nickname

_D, _T = limit.split("/")
limit = getTime(_T) + int(_D) * 1440
DB = dict()
fine = dict()

for i in range(N):
    curTime, partname, nickname = getLine()
    
    if not nickname in DB.keys():
        DB[nickname] = dict()
        DB[nickname][partname] = curTime
    else:
        if partname in DB[nickname].keys():
            prev = DB[nickname][partname]
            dist = max(0, curTime - prev - limit)
            if not nickname in fine.keys():
                fine[nickname] = 0
            fine[nickname] += dist * money
            del DB[nickname][partname]
        else:
            DB[nickname][partname] = curTime

ans = False
for nickname in sorted(fine.keys()):
    if fine[nickname] == 0:
        continue
    ans = True
    print(f"{nickname} {fine[nickname]}")

if ans == False:
    print(-1)